README.md
# gptpilot README

This is the README for your extension "gptpilot". After writing up a brief description, we recommend including the following sections.

## Features

Describe specific features of your extension including screenshots of your extension in action. Image paths are relative to this README file.

For example if there is an image subfolder under your extension project workspace:

\!\[feature X\]\(images/feature-x.png\)

> Tip: Many popular extensions utilize animations. This is an excellent way to show off your extension! We recommend short, focused animations that are easy to follow.

## Requirements

If you have any requirements or dependencies, add a section describing those and how to install and configure them.

## Extension Settings

Include if your extension adds any VS Code settings through the `contributes.configuration` extension point.

For example:

This extension contributes the following settings:

* `myExtension.enable`: Enable/disable this extension.
* `myExtension.thing`: Set to `blah` to do something.

## Known Issues

Calling out known issues can help limit users opening duplicate issues against your extension.

## Release Notes

Users appreciate release notes as you update your extension.

### 1.0.0

Initial release of ...

### 1.0.1

Fixed issue #.

### 1.1.0

Added features X, Y, and Z.

---

## Following extension guidelines

Ensure that you've read through the extensions guidelines and follow the best practices for creating your extension.

* [Extension Guidelines](https://code.visualstudio.com/api/references/extension-guidelines)

## Working with Markdown

You can author your README using Visual Studio Code. Here are some useful editor keyboard shortcuts:

* Split the editor (`Cmd+\` on macOS or `Ctrl+\` on Windows and Linux).
* Toggle preview (`Shift+Cmd+V` on macOS or `Shift+Ctrl+V` on Windows and Linux).
* Press `Ctrl+Space` (Windows, Linux, macOS) to see a list of Markdown snippets.

## For more information

* [Visual Studio Code's Markdown Support](http://code.visualstudio.com/docs/languages/markdown)
* [Markdown Syntax Reference](https://help.github.com/articles/markdown-basics/)

**Enjoy!**


CHANGELOG.md
# Change Log

All notable changes to the "gptpilot" extension will be documented in this file.

Check [Keep a Changelog](http://keepachangelog.com/) for recommendations on how to structure this file.

## [Unreleased]

- Initial release

src/extension.ts
import * as vscode from "vscode";
import axios from "axios";
import * as path from "path";
import * as fs from "fs";

export function activate(context: vscode.ExtensionContext) {

  let refactorDisposable = vscode.commands.registerCommand(
    "gptpilot.refactorCode",
    async () => {
      vscode.window.showInformationMessage("Refactoring code...");

      const editor = vscode.window.activeTextEditor;

      if (editor) {
        const selectedCode = editor.document.getText(editor.selection);
        const refactoredCode = await refactorCode(selectedCode);
        editor.edit((editBuilder) => {
			 editBuilder.replace(editor.selection, refactoredCode);
			});
      }
    }
  );

  let generateTestsDisposable = vscode.commands.registerCommand(
    "gptpilot.generateUnitTests",
    async () => {
      vscode.window.showInformationMessage("Generating unit tests...");
      const editor = vscode.window.activeTextEditor;
      try {
        if (editor) {
          const selectedCode = editor.document.getText(editor.selection);

          const editorPath = editor?.document.fileName;
          const posfix = editorPath?.split(".").pop();

          const unitTests = await generateUnitTests(selectedCode);

          const workspaceFolder = vscode.workspace.workspaceFolders[0].uri.fsPath;
          const fileName = `unitTest.${posfix}`;
          const filePath = path.join(workspaceFolder, fileName);

          // Convert the unit tests to a buffer before writing to the file
          const buffer = Buffer.from(unitTests);

          await vscode.workspace.fs.writeFile(
            vscode.Uri.file(filePath),
            buffer
          );

        }
      } catch (e) {
        vscode.window.showInformationMessage(
          "There is an error please try again"
        );

        throw e;
      }
    }
  );

  let checkVulnerability = vscode.commands.registerCommand(
    "gptpilot.checkVulnerability",

    async () => {
      vscode.window.showInformationMessage("Checking Vulnerability...");
      const editor = vscode.window.activeTextEditor;

      try {
        if (editor) {
          const selectedCode = editor.document.getText(editor.selection);

          // Call a function to generate the unit tests
          const vulnerabilities = await checkVulnerabiliity(selectedCode);

          const workspaceFolder = vscode.workspace.workspaceFolders[0].uri.fsPath;
          const fileName = `Vulnerabilities.txt`;
          const filePath = path.join(workspaceFolder, fileName);

          // Convert the unit tests to a buffer before writing to the file
          const buffer = Buffer.from(vulnerabilities);

          await vscode.workspace.fs.writeFile(vscode.Uri.file(filePath), buffer);
        }
      } catch (e) {
        vscode.window.showInformationMessage(
          "There is an error please try again"
        );
        throw e;
      }
    }
  );

  context.subscriptions.push(refactorDisposable);
  context.subscriptions.push(generateTestsDisposable);
  context.subscriptions.push(checkVulnerability);
}
async function refactorCode(selectedCode: string): Promise<string> {
  const apiKey = "API Key";
  const endpoint = "https://api.openai.com/v1/chat/completions";

  const prompt = `Respone only in code, Refactor the following code and make clean and simple:\n\n${selectedCode}\n`;

  try {
    const response = await axios.post(
      endpoint,
      {
        model: "gpt-3.5-turbo",
        messages: [{ role: "user", content: prompt }],
        max_tokens: selectedCode.length + 20,
        n: 1,
      },
      { headers: { Authorization: `Bearer ${apiKey}` } }
    );
    console.log(response.data.choices[0].message.content.trim());

    return response.data.choices[0].message.content.trim();
  } catch (e) {
    vscode.window.showInformationMessage("There is an error");
    throw e;
  }
}

async function generateUnitTests(selectedCode: string): Promise<string> {
  const apiKey = "API Key";
  const endpoint = "https://api.openai.com/v1/chat/completions";
  const editor = vscode.window.activeTextEditor;
  const fileName = editor?.document.fileName;
  const posfix = fileName?.split(".").pop();
  const prompt = `respone only in code and comments, Generate 3 unit tests in ${posfix} with all the needed imports and file configurations for the following code in the file ${fileName}: \n\n${selectedCode}\n `;
  console.log(`prompt name: ${prompt}`);
  try {
    5;
    const response = await axios.post(
      endpoint,
      {
        model: "gpt-3.5-turbo",
        messages: [{ role: "user", content: prompt }],
        n: 1,
        temperature: 0.1,
      },
      { headers: { Authorization: `Bearer ${apiKey}` } }
    );
    return response.data.choices[0].message.content.trim();
  } catch (e) {
    vscode.window.showInformationMessage("There is an error");
    throw e;
  }
}

async function checkVulnerabiliity(selectedCode: string): Promise<string> {
  vscode.window.showInformationMessage("Hello World from gpt-pilot!");

  const apiKey = "API Key";
  const endpoint = "https://api.openai.com/v1/chat/completions";
  const prompt = `detect vulnerabilities based on owasp standers without saying As an AI language model... on this code: \n\n${selectedCode}\n `;
  console.log(`prompt name: ${prompt}`);
  try {
    const response = await axios.post(
      endpoint,
      {
        model: "gpt-3.5-turbo",
        messages: [{ role: "user", content: prompt }],
        n: 1,
        temperature: 0.1,
        max_tokens: 250,
      },
      { headers: { Authorization: `Bearer ${apiKey}` } }
    );
    return response.data.choices[0].message.content.trim();
  } catch (e) {
    vscode.window.showInformationMessage("There is an error");
    throw e;
  }
}

export function deactivate() {}


out/extension.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.deactivate = exports.activate = void 0;
const vscode = require("vscode");
const axios_1 = require("axios");
const path = require("path");
function activate(context) {
    let refactorDisposable = vscode.commands.registerCommand("gptpilot.refactorCode", async () => {
        vscode.window.showInformationMessage("Refactoring code...");
        const editor = vscode.window.activeTextEditor;
        if (editor) {
            const selectedCode = editor.document.getText(editor.selection);
            const refactoredCode = await refactorCode(selectedCode);
            editor.edit((editBuilder) => {
                editBuilder.replace(editor.selection, refactoredCode);
            });
        }
    });
    let generateTestsDisposable = vscode.commands.registerCommand("gptpilot.generateUnitTests", async () => {
        vscode.window.showInformationMessage("Generating unit tests...");
        const editor = vscode.window.activeTextEditor;
        try {
            if (editor) {
                const selectedCode = editor.document.getText(editor.selection);
                const editorPath = editor?.document.fileName;
                const posfix = editorPath?.split(".").pop();
                const unitTests = await generateUnitTests(selectedCode);
                const workspaceFolder = vscode.workspace.workspaceFolders[0].uri.fsPath;
                const fileName = `unitTest.${posfix}`;
                const filePath = path.join(workspaceFolder, fileName);
                // Convert the unit tests to a buffer before writing to the file
                const buffer = Buffer.from(unitTests);
                await vscode.workspace.fs.writeFile(vscode.Uri.file(filePath), buffer);
            }
        }
        catch (e) {
            vscode.window.showInformationMessage("There is an error please try again");
            throw e;
        }
    });
    let checkVulnerability = vscode.commands.registerCommand("gptpilot.checkVulnerability", async () => {
        vscode.window.showInformationMessage("Checking Vulnerability...");
        const editor = vscode.window.activeTextEditor;
        try {
            if (editor) {
                const selectedCode = editor.document.getText(editor.selection);
                // Call a function to generate the unit tests
                const vulnerabilities = await checkVulnerabiliity(selectedCode);
                const workspaceFolder = vscode.workspace.workspaceFolders[0].uri.fsPath;
                const fileName = `Vulnerabilities.txt`;
                const filePath = path.join(workspaceFolder, fileName);
                // Convert the unit tests to a buffer before writing to the file
                const buffer = Buffer.from(vulnerabilities);
                await vscode.workspace.fs.writeFile(vscode.Uri.file(filePath), buffer);
            }
        }
        catch (e) {
            vscode.window.showInformationMessage("There is an error please try again");
            throw e;
        }
    });
    context.subscriptions.push(refactorDisposable);
    context.subscriptions.push(generateTestsDisposable);
    context.subscriptions.push(checkVulnerability);
}
exports.activate = activate;
async function refactorCode(selectedCode) {
    const apiKey = "API Key";
    const endpoint = "https://api.openai.com/v1/chat/completions";
    const prompt = `Respone only in code, Refactor the following code and make clean and simple:\n\n${selectedCode}\n`;
    try {
        const response = await axios_1.default.post(endpoint, {
            model: "gpt-3.5-turbo",
            messages: [{ role: "user", content: prompt }],
            max_tokens: selectedCode.length + 20,
            n: 1,
        }, { headers: { Authorization: `Bearer ${apiKey}` } });
        console.log(response.data.choices[0].message.content.trim());
        return response.data.choices[0].message.content.trim();
    }
    catch (e) {
        vscode.window.showInformationMessage("There is an error");
        throw e;
    }
}
async function generateUnitTests(selectedCode) {
    const apiKey = "API Key";
    const endpoint = "https://api.openai.com/v1/chat/completions";
    const editor = vscode.window.activeTextEditor;
    const fileName = editor?.document.fileName;
    const posfix = fileName?.split(".").pop();
    const prompt = `respone only in code and comments, Generate 3 unit tests in ${posfix} with all the needed imports and file configurations for the following code in the file ${fileName}: \n\n${selectedCode}\n `;
    console.log(`prompt name: ${prompt}`);
    try {
        5;
        const response = await axios_1.default.post(endpoint, {
            model: "gpt-3.5-turbo",
            messages: [{ role: "user", content: prompt }],
            n: 1,
            temperature: 0.1,
        }, { headers: { Authorization: `Bearer ${apiKey}` } });
        return response.data.choices[0].message.content.trim();
    }
    catch (e) {
        vscode.window.showInformationMessage("There is an error");
        throw e;
    }
}
async function checkVulnerabiliity(selectedCode) {
    vscode.window.showInformationMessage("Hello World from gpt-pilot!");
    const apiKey = "API Key";
    const endpoint = "https://api.openai.com/v1/chat/completions";
    const prompt = `detect vulnerabilities based on owasp standers without saying As an AI language model... on this code: \n\n${selectedCode}\n `;
    console.log(`prompt name: ${prompt}`);
    try {
        const response = await axios_1.default.post(endpoint, {
            model: "gpt-3.5-turbo",
            messages: [{ role: "user", content: prompt }],
            n: 1,
            temperature: 0.1,
            max_tokens: 250,
        }, { headers: { Authorization: `Bearer ${apiKey}` } });
        return response.data.choices[0].message.content.trim();
    }
    catch (e) {
        vscode.window.showInformationMessage("There is an error");
        throw e;
    }
}
function deactivate() { }
exports.deactivate = deactivate;
//# sourceMappingURL=extension.js.map